window.stop() should fire abort events on XMLHttpRequest asynchronously

cdumez opened the pull request:
#### 02e1d3e39f48975c3a6fcc396ad05574f2191759
<pre>
window.stop() should fire abort events on XMLHttpRequest asynchronously
<a href="https://bugs.webkit.org/show_bug.cgi?id=257397">https://bugs.webkit.org/show_bug.cgi?id=257397</a>

Reviewed by Darin Adler and Youenn Fablet.

window.stop() should fire an abort event on XMLHttpRequest asynchronously, to match
the behavior of Blink and Gecko.

It should schedule a asynchronous task, which should get cancelled if the
JavaScript calls `open()` on the XMLHttpRequest again. This is covered by WPT
tests.

* LayoutTests/imported/w3c/web-platform-tests/xhr/abort-after-stop.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/open-after-stop.window-expected.txt:
Rebaseline tests now that more checks are passing.

* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-basic-cors-not-enabled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-basic-cors-not-enabled.htm:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-cors-basic-setrequestheader-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-cors-basic-setrequestheader.htm:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-cors-setrequestheader-no-cred-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-cors-setrequestheader-no-cred.htm:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-network-error-sync-events.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-network-error-sync-events.sub.htm:
Fix tests to use get_host_info() instead of hardcoding custom domains which don&apos;t work with the
WebKit layout tests infrastructure. This improves test coverage. I&apos;ll upstream those changes.

* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::XMLHttpRequest):
(WebCore::XMLHttpRequest::open):
(WebCore::XMLHttpRequest::abort):
(WebCore::XMLHttpRequest::internalAbort):
(WebCore::XMLHttpRequest::abortError):
(WebCore::XMLHttpRequest::handleCancellation):
(WebCore::XMLHttpRequest::didFail):
* Source/WebCore/xml/XMLHttpRequest.h:
Align our behavior with Blink and Gecko.

Canonical link: <a href="https://commits.webkit.org/264765@main">https://commits.webkit.org/264765@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c67bf03591216d5a9ece7f0edcb57ca9ec1a34cc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8496 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8789 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/9062 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/10163 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8541 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/8505 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10779 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8740 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11410 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8643 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9684 "3 failures") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/7699 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/10318 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6982 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/7826 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15326 "19 flakes 123 failures") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/8103 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/9754 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/11278 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8395 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/6862 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7680 "Built successfully") | | | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2073 "Failed to push commit to Webkit repository") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11891 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/8140 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -1149,6 +1149,7 @@ void LocalDOMWindow::stop()
     if (!frame)
         return;
 
+    SetForScope isStopped { m_isStopping, true };
```
isStopping?
Also, shouldn't we need to write it as `m_isStopping, true, false`?

cdumez replied:
`isStopping` indeed, will fix.

With regards to your other comment, I don't think so. My version is a little safer in case of re-entrancy because it will restore whatever value `m_isStopping` had before (which is likely false but not necessarily in case we re-enter this function).

youennf commented on Source/WebCore/xml/XMLHttpRequest.cpp:
```diff
@@ -682,7 +681,7 @@ void XMLHttpRequest::abort()
     Ref<XMLHttpRequest> protectedThis(*this);
 
-    m_wasAbortedByClient = true;
+    m_wasAbortedByClient = false;
```
The web page is calling abort() so it seems `m_wasAbortedByClient` should be true.
If we change the meaning of `m_wasAbortedByClient`, we should change its name.
In this case, maybe we can remove it altogether.

cdumez replied:
You're right, not sure what happened here. I'll restore.

youennf commented on Source/WebCore/xml/XMLHttpRequest.cpp:
```diff
@@ -889,18 +889,28 @@ String XMLHttpRequest::statusText() const
+    }
+
     // The XHR specification says we should only fire an abort event if the cancelation was requested by the client.
     if (m_wasAbortedByClient && error.isCancellation()) {
```
With this patch, m_wasAbortedByClient is always false, except in this method, and it is the only place we check it. Can we use a local variable?

darinadler replied:
While it’s non-obvious, I studied the code and verified that @youennf is right about this. No other code is looking at `m_wasAbortedByClient`, and it will always be false entering this function, so we don’t need `m_wasAbortedByClient` any more.

cdumez replied:
It is set to true in abort().

darinadler replied:
But `abort` then unconditionally calls `immediateAbort`. That sets `m_error` unconditionally. So we will return early in that case and after the early return we are guaranteed that `m_wasAbortedByClient` is false. So I am sure we can get the same behavior without a data member.

cdumez replied:
Oh, I see. Thanks for clarifying. I'll look into it now.

youennf commented on Source/WebCore/xml/XMLHttpRequest.cpp:
```diff
@@ -889,18 +889,28 @@ String XMLHttpRequest::statusText() const
 }
 
-void XMLHttpRequest::didFail(const ResourceError& error)
+void XMLHttpRequest::handleCancellation(const ResourceError&)
```
Why having a `const ResourceError&`?

cdumez replied:
Will drop.

youennf reviewed (approved):
r=me.
Please land the WPT PR first though.

darinadler commented on Source/WebCore/xml/XMLHttpRequest.cpp:
```diff
@@ -135,8 +135,7 @@ XMLHttpRequest::~XMLHttpRequest()
 {
-    ASSERT(scriptExecutionContext());
-    return downcast<Document>(scriptExecutionContext());
+    return dynamicDowncast<Document>(scriptExecutionContext());
```
I don’t understand why this is `dynamicDowncast`. It doesn’t need to be just because `scriptExecutionContext` might be `nullptr`. Can it be a non-document context? I suppose it’s safe just in case, but is there any reason that case exists?

cdumez replied:
XMLHttpRequest is exposed to workers:
```
[
    ActiveDOMObject,
    ExportMacro=WEBCORE_EXPORT,
    Exposed=(Window,DedicatedWorker,SharedWorker),
    JSCustomMarkFunction,
    JSGenerateToJSObject,
    JSGenerateToNativeObject,
    ReportExtraMemoryCost,
] interface XMLHttpRequest : XMLHttpRequestEventTarget {
```

As a result, this function was really risky. The caller has to be certain they were not on a worker thread before calling this. In my patch, I wanted access to the window (which I get from the document) and I felt it was more convenient to do a null check than explicitly making sure we're not on a worker thread.

annevk commented on LayoutTests/imported/w3c/web-platform-tests/xhr/send-network-error-sync-events.sub.htm:
```diff
@@ -17,7 +17,7 @@
             var xhr = new XMLHttpRequest();
 
-            xhr.open("POST", "http://nonexistent.{{host}}:{{ports[http][0]}}", false);
+            xhr.open("POST", "http://{{host}}:3", false); // Bad port.
```
3 is not a bad port as far as I can tell from https://fetch.spec.whatwg.org/#bad-port.

cdumez replied:
Thanks, fixed!

cdumez committed:
window.stop() should fire abort events on XMLHttpRequest asynchronously
https://bugs.webkit.org/show_bug.cgi?id=257397

Reviewed by Darin Adler and Youenn Fablet.

window.stop() should fire an abort event on XMLHttpRequest asynchronously, to match
the behavior of Blink and Gecko.

It should schedule a asynchronous task, which should get cancelled if the
JavaScript calls `open()` on the XMLHttpRequest again. This is covered by WPT
tests.

* LayoutTests/imported/w3c/web-platform-tests/xhr/abort-after-stop.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/open-after-stop.window-expected.txt:
Rebaseline tests now that more checks are passing.

* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-basic-cors-not-enabled-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-basic-cors-not-enabled.htm:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-cors-basic-setrequestheader-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-cors-basic-setrequestheader.htm:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-cors-setrequestheader-no-cred-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-authentication-cors-setrequestheader-no-cred.htm:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-network-error-sync-events.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-network-error-sync-events.sub.htm:
Fix tests to use get_host_info() instead of hardcoding custom domains which don't work with the
WebKit layout tests infrastructure. This improves test coverage. I'll upstream those changes.

* Source/WebCore/xml/XMLHttpRequest.cpp:
(WebCore::XMLHttpRequest::XMLHttpRequest):
(WebCore::XMLHttpRequest::open):
(WebCore::XMLHttpRequest::abort):
(WebCore::XMLHttpRequest::internalAbort):
(WebCore::XMLHttpRequest::abortError):
(WebCore::XMLHttpRequest::handleCancellation):
(WebCore::XMLHttpRequest::didFail):
* Source/WebCore/xml/XMLHttpRequest.h:
Align our behavior with Blink and Gecko.

Canonical link: https://commits.webkit.org/264765@main
